Modules/Server/Sources/PocketCastsServer/Public/Cache/EpisodeInfoHandler.swift

```diff
@@ -0,0 +1,119 @@
+    }
+
+    struct ShowNotes: Decodable {
+        let podcast: ShowNotesPodcast
+    }
+
+    struct ShowNotesPodcast: Decodable {
+        let episodes: [ShowNotesEpisode]
+    }
+
+    struct ShowNotesEpisode: Decodable {
+        let uuid: String
+        let showNotes: String
+        let image: String?
+    }
+
+    public func loadShowNotes(podcastUuid: String, episodeUuid: String, cached: ((String) -> Void)? = nil, completion: ((String?) -> Void)?) {
+        var cachedNotes = ""
+        var didSendCachedNotes = false
+
+        requestShowNotes(for: podcastUuid, cached: { showNotes in
+            if let notes = showNotes?.episodes.first(where: { $0.uuid == episodeUuid })?.showNotes {
+                cached?(notes)
+                cachedNotes = notes
+                didSendCachedNotes = true
+            }
+        }, completion: { showNotes in
+            if let episodeNotes = showNotes?.episodes.first(where: { $0.uuid == episodeUuid })?.showNotes {
+                if didSendCachedNotes, episodeNotes == cachedNotes {
+                    return
+                }
+
+                completion?(episodeNotes)
+            } else if !didSendCachedNotes {
+                completion?(CacheServerHandler.noShowNotesMessage)
+            }
+        })
+    }
+
+    /// Multiple calls can be made to this method and we'll ensure that only
+    /// one request per time is made to the show notes endpoint — avoiding
+    /// multiple calls.
+    /// - Parameters:
+    ///   - for: a podcast UUID
+    ///   - cached: a closure that receive a cached notes information
+    ///   - completion: a closure that *might* receive a cached or update notes information
+    private func requestShowNotes(for podcastUuid: String, cached: @escaping (ShowNotesPodcast?) -> Void, completion: @escaping (ShowNotesPodcast?) -> Void) {
+        showNotesCachedCompletionBlocks.append(cached)
+        showNotesCompletionBlocks.append(completion)
+
+        guard !requestingNotes else { return }
+
+        requestingNotes = true
+
+        let decoder = JSONDecoder()
+        decoder.keyDecodingStrategy = .convertFromSnakeCase
+
+        let url = ServerHelper.asUrl(ServerConstants.Urls.cache() + "mobile/show_notes/full/\(podcastUuid)")
+        var request = URLRequest(url: url)
+        request.cachePolicy = .reloadRevalidatingCacheData
+
+        // Check for any cached version and if there's any call the cached completion blocks
+        if let cachedResponse = showNotesUrlCache.cachedResponse(for: request),
+           let showNotes = try? decoder.decode(ShowNotes.self, from: cachedResponse.data) {
+
+            showNotesCachedCompletionBlocks.forEach { $0(showNotes.podcast) }
+            showNotesCachedCompletionBlocks = []
+            requestingNotes = false
+        }
+
+        // Call the endpoint to request for a more up-to-date notes information
+        URLSession.shared.dataTask(with: request) { [weak self] data, response, _ in
```

Comment:
Can we switch this to the async version since it's much cleaner?

Reply:
Do you mean the async `URLSession.shared.data`? Or `JSONDecodableURLTask`? I give a try with the async `URLSession.shared.data` and I thought that the end result was much more confusing:

```
Task {
            // Call the endpoint to request for a more up-to-date notes information
            do {
                let (data, response) = try await URLSession.shared.data(for: request)

                let responseToCache = CachedURLResponse(response: response, data: data)
                showNotesUrlCache.storeCachedResponse(responseToCache, for: request)

                guard let showNotes = try? decoder.decode(ShowNotes.self, from: data) else {
                    showNotesCompletionBlocks.forEach { $0(nil) }
                    showNotesCompletionBlocks = []
                    return
                }

                showNotesCompletionBlocks.forEach { $0(showNotes.podcast) }
                showNotesCompletionBlocks = []
            } catch {
                showNotesCompletionBlocks.forEach { $0(nil) }
                showNotesCompletionBlocks = []
            }

            requestingNotes = false
        }
```

Alternatively, I can add a new method to `JSONDecodableURLTask` to return the entity **and** the request result, but I wonder if this clashes with the single responsibility of `JSONDecodableURLTask`.

Maybe I'm missing something tho! Thoughts?